openni/plugin.cpp

```diff
@@ -0,0 +1,160 @@
+#include "../common/data_format.hpp"
+#include "../common/phonebook.hpp"
+#include "../common/pose_prediction.hpp"
+#include "../common/relative_clock.hpp"
+#include "../common/switchboard.hpp"
+#include "../common/threadloop.hpp"
+
+#include <iomanip>
+#include <opencv/cv.hpp>
+#include <opencv2/core/core.hpp>
+#include <opencv2/highgui/highgui.hpp>
+#include <OpenNI.h>
+
+using namespace ILLIXR;
+
+class openni_plugin : public ILLIXR::threadloop {
+public:
+    openni_plugin(std::string name_, phonebook* pb_)
+        : threadloop{name_, pb_}
+        , sb{pb->lookup_impl<switchboard>()}
+        , _m_clock{pb->lookup_impl<RelativeClock>()}
+        , _m_rgb_depth{sb->get_writer<rgb_depth_type>("rgb_depth")} {
+        if (!camera_initialize()) {
+            std::cout << "Initialization failed" << std::endl;
+            exit(0);
+        }
+    }
+
+    virtual ~openni_plugin() override {
+        color.destroy();
+        depth.destroy();
+    }
+
+protected:
+    virtual skip_option _p_should_skip() override {
+        auto now = std::chrono::steady_clock::now();
+        cam_time = std::chrono::time_point_cast<std::chrono::milliseconds>(now).time_since_epoch().count();
+        if (cam_time > last_ts) {
+            std::this_thread::sleep_for(std::chrono::milliseconds{2});
+            return skip_option::run;
+        } else {
+            return skip_option::skip_and_yield;
+        }
+    }
+
+    virtual void _p_one_iteration() override {
+        RAC_ERRNO_MSG("openni at start of _p_one_iteration");
+
+        // read cam data
+        color.readFrame(&colorFrame);
+        depth.readFrame(&depthFrame);
+        // get timestamp
+        assert(colorFrame.getTimestamp() != depthFrame.getTimestamp());
+
+        // convert to cv format
+        cv::Mat colorMat;
+        colorMat.create(colorFrame.getHeight(), colorFrame.getWidth(), CV_8UC3);
+        const openni::RGB888Pixel* colorBuffer = (const openni::RGB888Pixel*) colorFrame.getData();
+        memcpy(colorMat.data, colorBuffer, 3 * colorFrame.getHeight() * colorFrame.getWidth() * sizeof(uint8_t));
+        cv::cvtColor(colorMat, colorMat, cv::COLOR_BGR2BGRA);
+        colorMat.convertTo(colorMat, CV_8UC4);
+
+        cv::Mat depthMat;
+        depthMat.create(depthFrame.getHeight(), depthFrame.getWidth(), CV_16UC1);
+        const openni::DepthPixel* depthBuffer = (const openni::DepthPixel*) depthFrame.getData();
+        memcpy(depthMat.data, depthBuffer, depthFrame.getHeight() * depthFrame.getWidth() * sizeof(uint16_t));
```

Comment:
What is the unit of captured depth image?
The loaded depth pixel should be converted to meters or millimeters instead of the raw values.
Probably want to run that conversion before publish the depth image.

Reply:
Have verified that OpenNI publishes depth image in mm.